runtime/auth/aws-credentials/common/src/aws/smithy/kotlin/runtime/auth/awscredentials/Credentials.kt

```diff
@@ -7,26 +7,82 @@ package aws.smithy.kotlin.runtime.auth.awscredentials
  */
-// FIXME - should probably be an interface
-public data class Credentials(
-    val accessKeyId: String,
-    val secretAccessKey: String,
-    val sessionToken: String? = null,
+public interface Credentials : Identity {
+    /**
+     * Identifies the user interacting with services
+     */
+    public val accessKeyId: String
+
+    /**
+     * Secret key used to authenticate the user and sign requests
+     */
+    public val secretAccessKey: String
+
+    /**
+     * Session token associated with short term credentials with an expiration.
+     */
+    public val sessionToken: String?
+        get() = null
+
+    /**
+     * The name of the credentials provider that sourced these credentials (if known).
+     */
+    public val providerName: String?
+        get() = attributes.getOrNull(IdentityAttributes.ProviderName)
+}
+
+public fun Credentials.copy(
+    accessKeyId: String = this.accessKeyId,
+    secretAccessKey: String = this.secretAccessKey,
+    sessionToken: String? = this.sessionToken,
+    expiration: Instant? = this.expiration,
+    providerName: String? = this.providerName,
+    attributes: Attributes? = this.attributes,
+): Credentials {
+    val updatedAttributes = if (this.providerName != null && providerName == null) {
+        // if provider name was previously set and is updated to null we need to remove it from the current attributes
+        attributes?.toMutableAttributes()?.apply {
+            remove(IdentityAttributes.ProviderName)
+        }?.takeIf(Attributes::isNotEmpty)
+    } else {
+        attributes
+    }
+    return Credentials(accessKeyId, secretAccessKey, sessionToken, expiration, providerName, updatedAttributes)
+}
+
+private data class CredentialsImpl(
+    override val accessKeyId: String,
+    override val secretAccessKey: String,
+    override val sessionToken: String? = null,
     override val expiration: Instant? = null,
-    val providerName: String? = null,
-) : Identity {
-    override val attributes: MutableAttributes = mutableAttributes()
-    init {
-        providerName?.let {
-            attributes[IdentityAttributes.ProviderName] = it
-        }
+    override val attributes: Attributes = emptyAttributes(),
+) : Credentials
+
+/**
+ * Create a new [Credentials] instance
+ */
+public fun Credentials(
+    accessKeyId: String,
+    secretAccessKey: String,
+    sessionToken: String? = null,
+    expiration: Instant? = null,
+    providerName: String? = null,
+    attributes: Attributes? = null,
+): Credentials {
+    val resolvedAttributes = if (providerName != null && attributes?.getOrNull(IdentityAttributes.ProviderName) != providerName) {
+        val merged = attributes?.toMutableAttributes() ?: mutableAttributes()
+        merged.setIfValueNotNull(IdentityAttributes.ProviderName, providerName)
+        merged
+    } else {
+        attributes ?: emptyAttributes()
     }
+
+    return CredentialsImpl(accessKeyId, secretAccessKey, sessionToken, expiration, resolvedAttributes)
 }
```

Comment:
**Question**: Any reason not to make this an `invoke` function on `Credentials.Companion`?